Add patchwork interface

rsarky opened the pull request:
This PR adds supoort to pull patches from patchwork.
There are 2 ways PaStA can pull patches:
- A raw mailbox containing patches from patchwork. (can be used for initial import)
- Through Patchwork's API. In this case Pasta will only import patches that it doesnt have already.

metp committed:
Add python-requests as dependency

Used for HTTP requests to Patchwork.

Signed-off-by: Mete Polat <metepolat2000@gmail.com>

rralf commented:
Thanks! I'll have a look at it tomorrow.

rralf commented:
One thing: Could you take the adjustments of Config.py and load/write_index to a separate commits?

rsarky commented:
> One thing: Could you take the adjustments of Config.py and load/write_index to a separate commits?

Sure

rsarky committed:
pypasta: feat: allow pasta to pull patches from Patchwork

Patchwork is a new mail storage container that inherits from MailContainer. It
is used to pull patches from patchwork instances.

Usage:
Define a mbox.patchwork section in the configuration:
...
[mbox.patchwork]
url = <patchwork-url>
projects = [{id = x, initial_archive = "path/to/archive", list_email="list@domain.org"}, ...]
page_size = 30
...

If initial_archive is defined, Pasta will pull patches from the initial archive
treating it as a raw mailbox.

If initial_archive is not set, PaStA will pull patches from Patchwork through
its API, with respect to patchwork's pagination. We will only pull patches
that are not already in PaStA.

Signed-off-by: Rohit Sarkar <rohitsarkar5398@gmail.com>
[ralf: massage commit message, streamline things]

rsarky committed:
docs: add documentation for set up with Patchwork

Signed-off-by: Rohit Sarkar <rohitsarkar5398@gmail.com>
[ralf: update documentation]

rsarky commented:
Updated

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -392,6 +492,20 @@ def __init__(self, config):
+                for message_id in project.get_ids():
+                    self.add_mail_to_list(message_id, listaddr)
+                self.patchwork_projects.append(project)
+        else:
```
python complains about indentation here

rralf replied:
I guess we can simply remove the else statement, right?

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -486,7 +605,10 @@ def get_raws(self, message_id):
+        if len(self.patchwork_projects) > 0:
+            for project in self.patchwork_projects:
+                ids |= project.get_ids(time_window)
+        else:
```
same here, incorrect indentation

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -475,6 +589,11 @@ def get_messages(self, id):
+            for project in self.patchwork_projects:
+                if message_id in project:
+                    raws += project[message_id]
+        else:
```
and here, my interpreter doesn't like it.

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -475,6 +589,11 @@ def get_messages(self, id):
     def get_raws(self, message_id):
         raws = list()
 
+        if len(self.patchwork_projects) > 0:
```
We don't need this check, the for loop will do it already for us

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -486,7 +605,10 @@ def get_raws(self, message_id):
     def get_ids(self, time_window=None, allow_invalid=False, lists=None):
         ids = set()
-
+        if len(self.patchwork_projects) > 0:
```
same here, the for loop will take care of this check.

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -501,6 +623,10 @@ def get_ids(self, time_window=None, allow_invalid=False, lists=None):
+        if len(self.patchwork_projects) > 0:
+            for project in self.patchwork_projects:
+                project.update()
+        else:
```
indentation.

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -501,6 +623,10 @@ def get_ids(self, time_window=None, allow_invalid=False, lists=None):
         return ids
 
     def update(self):
+        if len(self.patchwork_projects) > 0:
```
remove the check.

rralf reviewed (changes requested):
Hi,

before I continue with my review: please remove unnecessary if len() checks, and take care of the indentation. In all cases, you can remove the else statement. While it's surely not a good idea, there's no need to disallow raw mboxes in combination with patchwork. It's not required for our use-case, but there's no reason to forbid it.

rralf reviewed (changes requested):
This patch breaks PaStA if we don't use a patchwork configuration:

  File "/media/Datenfass/PaStA/pypasta/Config.py", line 199, in __init__
    mbox_patchwork = mbox['patchwork']
KeyError: 'patchwork'

Please take care to not break the system. Only address patchwork-related config items, if a patchwork configuration is being used.

rralf reviewed (changes requested):
Please split it up into two patches: The first one to change load_index, the second one that pulls out the method.

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -179,28 +179,51 @@ def load_file(filename, must_exist=True):
 class MailContainer:
     @staticmethod
-    def load_index(f_index):
+    def load_index(f_index, is_patchwork_index=False):
```
TBH, I don't like this approach. We're calling the method like

load_index('/filename', True)

No one knows what "True" or "False" means on caller-site, if we use it as an argument. And I'm sure we don't even need this argument.

rsarky replied:
Since it's a keyword argument, the caller will have to include the keyword ie `is_patchwork_index` while making the call. Thus making it explicit.

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -179,28 +179,51 @@ def load_file(filename, must_exist=True):
+            if is_patchwork_index:
+                index[message_id].append((dtime, date, location, entry[3]))
+            else:
+                index[message_id].append((dtime, date, location))
```
One can only guess what entry[3] means, and know we have a rather complex differentiation, depending on is_patchwork_index.

rralf replied:
Let me propose a different approach, will get back to you.

rralf replied:
Please have a look at the latest two commits on next:
https://github.com/lfd/PaStA/commit/fc11490ef6ab5a7795c2d9d3d3ead12be8a892db and https://github.com/lfd/PaStA/commit/ae26e9be8713da08311ad9040bb8a23a65b3794e .

Should work with public inboxes as well as the patchwork backend. The patchwork-id will only be loaded/written if present. What do you think?

rsarky commented:
> Hi,
> 
> before I continue with my review: please remove unnecessary if len() checks, and take care of the indentation. In all cases, you can remove the else statement. While it's surely not a good idea, there's no need to disallow raw mboxes in combination with patchwork. It's not required for our use-case, but there's no reason to forbid it.

In all the cases, the else and the conditions based on len were introduced to disallow PaStA using Patchwork and public or raw mailboxes simultaneously. If we do want them to function together, I guess I can remove those checks

rralf commented:
> > Hi,
> > before I continue with my review: please remove unnecessary if len() checks, and take care of the indentation. In all cases, you can remove the else statement. While it's surely not a good idea, there's no need to disallow raw mboxes in combination with patchwork. It's not required for our use-case, but there's no reason to forbid it.
> 
> In all the cases, the else and the conditions based on len were introduced to disallow PaStA using Patchwork and public or raw mailboxes simultaneously. If we do want them to function together, I guess I can remove those checks

Yep, let's remove them.

rsarky commented:
> This patch breaks PaStA if we don't use a patchwork configuration:
> 
> File "/media/Datenfass/PaStA/pypasta/Config.py", line 199, in **init**
> mbox_patchwork = mbox['patchwork']
> KeyError: 'patchwork'
> 
> Please take care to not break the system. Only address patchwork-related config items, if a patchwork configuration is being used.

The same thing would happen if we removed `[mbox.raw]` or `[mbox.pubin]` from the Config. This PR assumes that the config will by default contain the line `[mbox.patchwork]`, this is the same way as pubins or raw mboxes are handled

rralf commented:
> > This patch breaks PaStA if we don't use a patchwork configuration:
> > File "/media/Datenfass/PaStA/pypasta/Config.py", line 199, in **init**
> > mbox_patchwork = mbox['patchwork']
> > KeyError: 'patchwork'
> > Please take care to not break the system. Only address patchwork-related config items, if a patchwork configuration is being used.
> 
> The same thing would happen if we removed `[mbox.raw]` or `[mbox.pubin]` from the Config. This PR assumes that the config will by default contain the line `[mbox.patchwork]`, this is the same way as pubins or raw mboxes are handled

Oh, does it? Thanks for the bug report! :+1: 
Will fix this...

rsarky commented:
> Please split it up into two patches: The first one to change load_index, the second one that pulls out the method.


I didnt pull out the `load_index` method. That was done by you in a previous commit.
Are you talking about the `load_mbox_file` method by any chance?

rsarky commented:
I wonder why the indentation is breaking. I used to use vim, but now I switched to vscode that uses pep8 style for formatting

rsarky commented:
> Hi,
> 
> before I continue with my review: please remove unnecessary if len() checks, and take care of the indentation. In all cases, you can remove the else statement. While it's surely not a good idea, there's no need to disallow raw mboxes in combination with patchwork. It's not required for our use-case, but there's no reason to forbid it.

Done

rralf commented:
> > > This patch breaks PaStA if we don't use a patchwork configuration:
> > > File "/media/Datenfass/PaStA/pypasta/Config.py", line 199, in **init**
> > > mbox_patchwork = mbox['patchwork']
> > > KeyError: 'patchwork'
> > > Please take care to not break the system. Only address patchwork-related config items, if a patchwork configuration is being used.
> > 
> > 
> > The same thing would happen if we removed `[mbox.raw]` or `[mbox.pubin]` from the Config. This PR assumes that the config will by default contain the line `[mbox.patchwork]`, this is the same way as pubins or raw mboxes are handled
> 
> Oh, does it? Thanks for the bug report! +1
> Will fix this...

Aah, we do not have a bug here. We set default values in resource's common.cfg:

https://github.com/lfd/PaStA-resources/blob/master/common/default.cfg

So the fix is simple: In future, we must add an empty [mbox.patchwork] entry to default.cfg. But we can postpone that for the moment.

rsarky commented:
> > > > This patch breaks PaStA if we don't use a patchwork configuration:
> > > > File "/media/Datenfass/PaStA/pypasta/Config.py", line 199, in **init**
> > > > mbox_patchwork = mbox['patchwork']
> > > > KeyError: 'patchwork'
> > > > Please take care to not break the system. Only address patchwork-related config items, if a patchwork configuration is being used.
> > > 
> > > 
> > > The same thing would happen if we removed `[mbox.raw]` or `[mbox.pubin]` from the Config. This PR assumes that the config will by default contain the line `[mbox.patchwork]`, this is the same way as pubins or raw mboxes are handled
> > 
> > 
> > Oh, does it? Thanks for the bug report! +1
> > Will fix this...
> 
> Aah, we do not have a bug here. We set default values in resource's common.cfg:
> 
> https://github.com/lfd/PaStA-resources/blob/master/common/default.cfg
> 
> So the fix is simple: In future, we must add an empty [mbox.patchwork] entry to default.cfg. But we can postpone that for the moment.

Yup makes sense

rralf commented:
Rebased your work and pushed it to wip-patchwork-integration. Will test it now.

rralf commented on README.md:
```diff
@@ -248,3 +248,47 @@ To compare all mails on the list against upstream:
+[mbox]
+...
+
+USE_PATCHWORK_ID = true
```
no longer required

rsarky replied:
> We should allow to have no initial_archive set in the configuration file.

So do you mean instead of checking for empty string in `initial_archive` property, we let the user completely remove that key?

It seems like the another way to do the same thing. Are there any advantages to remove the key instead of setting it to an empty string?

rralf replied:
Usually people comment out configuration options that they don't need. This is what I intuitively tried. I commented out initial_archive after the first run. Resulted in a crash of PaStA.

rsarky replied:
initial_archive key is associated with a projectid and list email. ie it is part of the same object.
So I dont think that can be commented out.

If we keep a separate array of initial_archives and separate array of projects that might be troublesome as both the arrays may have conflicting information

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -392,6 +492,20 @@ def __init__(self, config):
+        self.patchwork_projects = []
+        if len(config.mbox_patchwork_projects):
+            log.info('Loading Patchwork projects...')
+            for project_id, f_mbox_raw, listaddr in config.mbox_patchwork_projects:
```
The loop can go down one indentation level

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -392,6 +492,20 @@ def __init__(self, config):
+            for project_id, f_mbox_raw, listaddr in config.mbox_patchwork_projects:
+                self.lists.add(listaddr)
+                f_index = os.path.join(
+                    config.d_mbox, 'index', 'patchwork.%u' % project_id)
```
Yep, looks good!

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -392,6 +492,20 @@ def __init__(self, config):
+                for message_id in project.get_ids():
+                    self.add_mail_to_list(message_id, listaddr)
+                self.patchwork_projects.append(project)
+
```
Hmm. I can see a lot of duplicated startup code for all three different mailboxes. Maybe we can somehow consolidate the startup. But let's do that later.

rsarky replied:
Yes, that's a good idea

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -371,6 +375,102 @@ def __getitem__(self, message_id):
+        return self.page_iterator(resp, yield_value, since)
+
+    def update(self):
+        if self.f_mbox_raw != "":
```
Can be simplified to:
if self.f_mbox_raw:

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -16,10 +16,14 @@
 from logging import getLogger
 from subprocess import call
+from urllib.parse import urljoin
+from datetime import datetime
```
Hmm, above, we already import datetime. Importing it this way requires to maintain existing code.

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -371,6 +375,102 @@ def __getitem__(self, message_id):
+            log.info('Processing raw mailbox for Patchwork project id %u' %
+                     self.project_id)
+            load_mbox_file(self.f_mbox_raw, str(self.project_id),
+                           self.d_mbox, mode='patchwork')
```
Add a report here that the initial import was successful, and guide the user to remove the initial archive configuration option. Return here.

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -371,6 +375,102 @@ def __getitem__(self, message_id):
+                     self.project_id)
+            load_mbox_file(self.f_mbox_raw, str(self.project_id),
+                           self.d_mbox, mode='patchwork')
+        else:
```
And remove the else case, and put the indentation one level down.

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -371,6 +375,102 @@ def __getitem__(self, message_id):
+            greatest_patchwork_id = None
+            if len(self.index) > 0:
+                greatest_patchwork_id = max(
+                    [entry[3] for entries in self.index.values() for entry in entries])
```
Shouldn't we start from greatest_patchwork_id + 1? Otherwise, we will create duplicate entries for the highest patchwork id.

rsarky replied:
No. You can check the code in the `page_iterator` method, we "break" out of the generator when we encounter the patch with the highest patch id without yielding it.

rralf reviewed (changes requested):
Hi,

just successfully run your script. There are still some issues, we receive duplicate messages. Let's talk about that in a short video conference. Will send you an email tomorrow.

Thanks
  Ralf

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -371,6 +375,102 @@ def __getitem__(self, message_id):
+            patches = self.download_patches(since=greatest_patchwork_id)
+            pulled = 0
+            for patch in patches:
+                date, message_id, message_text, patchwork_id = patch
```
Hmm. We need to either get the API under control, or check if the patchwork-id is already present.

rsarky replied:
We can do some empirical tests to see if the API returns a patch with the same patchwork id twice, but I dont think this should happen. I am not sure if this is what you mean

rsarky commented:
> Hi,
> 
> just successfully run your script. There are still some issues, we receive duplicate messages. Let's talk about that in a short video conference. Will send you an email tomorrow.
> 
> Thanks
> Ralf

Sure

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -374,6 +377,100 @@ def __getitem__(self, message_id):
+        pages = self.download_patches()
+        pulled = 0
+        for index, page in enumerate(pages):
+            for patch in page:
```
This loop implicitly calls the iterator. The iterator does IO, and can fail. Please don't combine iterators with side-effects and IO.

rralf replied:
Other than that, we download all patches of the page, which is not necessary, as we should only load the ones that we're missing.

rralf replied:
Will return with a patch that implements that.

rsarky commented:
Added some error handling. Tested it out by stopping the patchwork server while the update was running. Works!

Added it to [this](https://github.com/lfd/PaStA/pull/57/commits/bd498e60542286cbe028f85fad3cbed702c98e60) commit

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -374,6 +377,122 @@ def __getitem__(self, message_id):
+        error = False
+        while next_page and not error:
+            log.info('Querying page %u' % next_page)
+            try:
```
We can do it simpler: Use one big try block for the while loop. It doesn't matter if the error occured in _get_page or _pull_patch. We have to stop in any case.

rsarky replied:
Hmm, makes sense

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -374,6 +377,122 @@ def __getitem__(self, message_id):
+                     for idx_entry in self.index.values()])
+        next_page = len(patchwork_ids)//self.page_size + 1
+        pulled = 0
+        error = False
```
try:
[...]

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -374,6 +377,122 @@ def __getitem__(self, message_id):
+            log.info('Querying page %u' % next_page)
+            try:
+                next_page, page = self._get_page(next_page)
+            except requests.exceptions.RequestException:
```
That's not sufficient. I just got:
urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))

RequestException is only one possible error. We need wider exception handling.

rsarky replied:
Seems like that was a problem with the url, not the http requests.
Anyway, seems like I will introduce a "catch all" then

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -374,6 +377,122 @@ def __getitem__(self, message_id):
+            try:
+                next_page, page = self._get_page(next_page)
+            except requests.exceptions.RequestException:
+                log.info(' Error fetching page %u. Stopping update' % next_page)
```
log.error, as it's an error.

No need to print next_page, we just printed it in line 446. Better print the actual cause of the eception

catch Exception as e:
  log.error(' An error occured during communication with patchwork:')
  log.error(' %s' % str(e))

rralf commented:
Looks good now! Thanks, merged to next and wip-patchwork-integration.
